Model/Service/Index.php

```diff
@@ -0,0 +1,286 @@
+     * @param NostoHelperScope $nostoHelperScope
+     * @param NostoLogger $logger
+     * @param NostoIndexCollectionFactory $nostoIndexCollectionFactory
+     * @param TimezoneInterface $magentoTimeZone
+     * @param NostoDataHelper $nostoDataHelper
+     * @param Sync $nostoSyncService
+     */
+    public function __construct(
+        IndexRepository $indexRepository,
+        Builder $indexBuilder,
+        ProductRepository $productRepository,
+        NostoProductBuilder $nostoProductBuilder,
+        NostoHelperScope $nostoHelperScope,
+        NostoLogger $logger,
+        NostoIndexCollectionFactory $nostoIndexCollectionFactory,
+        TimezoneInterface $magentoTimeZone,
+        NostoDataHelper $nostoDataHelper,
+        Sync $nostoSyncService
+    ) {
+        parent::__construct($nostoDataHelper, $logger);
+        $this->indexRepository = $indexRepository;
+        $this->indexBuilder = $indexBuilder;
+        $this->productRepository = $productRepository;
+        $this->nostoProductBuilder = $nostoProductBuilder;
+        $this->nostoHelperScope = $nostoHelperScope;
+        $this->nostoIndexCollectionFactory = $nostoIndexCollectionFactory;
+        $this->magentoTimeZone = $magentoTimeZone;
+        $this->nostoSyncService = $nostoSyncService;
+    }
+
+    /**
+     * Handles only the first step of indexing
+     * Create one if row does not exits
+     * Else set row to dirty
+     *
+     * @param ProductCollection $collection
+     * @param Store $store
+     * @throws NostoException
+     */
+    public function invalidateOrCreate(ProductCollection $collection, Store $store)
+    {
+        $this->startBenchmark(
+            self::BENCHMARK_NAME_INVALIDATE,
+            self::BENCHMARK_BREAKPOINT_INVALIDATE
+        );
+        $collection->setPageSize(self::PRODUCT_DATA_BATCH_SIZE);
+        $iterator = new Iterator($collection);
+        $iterator->each(function (Product $item) use ($store) {
+            $this->updateOrCreateDirtyEntity($item, $store);
+            $this->tickBenchmark(self::BENCHMARK_NAME_INVALIDATE);
+        });
+        $this->logBenchmarkSummary(self::BENCHMARK_NAME_INVALIDATE, $store);
+    }
+
+    /**
+     * @param Product $product
+     * @param Store $store
+     * @return ProductIndexInterface|null
+     */
+    public function updateOrCreateDirtyEntity(Product $product, Store $store)
+    {
+        $indexedProduct = $this->indexRepository->getByProductIdAndStoreId($product->getId(), $store->getId());
+        try {
+            if ($indexedProduct instanceof ProductIndexInterface) {
+                $indexedProduct->setIsDirty(true);
+                $indexedProduct->setUpdatedAt($this->magentoTimeZone->date());
+            } else {
+                /* @var Product $fullProduct */
+                $fullProduct = $this->loadMagentoProduct($product->getId(), $store->getId());
+                $indexedProduct = $this->indexBuilder->build($fullProduct, $store);
```

Comment:
This class should not have ANY logic for building a product. It should simply create an empty row if there wasn't one. It is the responsibility of the second indexer to actually build the product.